File: src/main/java/de/tlmrgvf/drtd/dsp/component/BitConverter.java
/*
 *
 * BSD 2-Clause License
 *
 * Copyright (c) 2020, Till Mayer
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package de.tlmrgvf.drtd.dsp.component;

import de.tlmrgvf.drtd.Drtd;
import de.tlmrgvf.drtd.dsp.PipelineComponent;
import de.tlmrgvf.drtd.utils.Utils;
import de.tlmrgvf.drtd.utils.structure.GenericRingBuffer;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.util.logging.Logger;

public final class BitConverter extends PipelineComponent<Boolean, Boolean> {
    private static BufferedImage IMAGE;
    private static Dimension SIZE;

    static {
        try {
            IMAGE = ImageIO.read(Drtd.readResourceStream("bitconverter.png"));
            SIZE = new Dimension(IMAGE.getWidth(), IMAGE.getHeight());
        } catch (IOException e) {
            Drtd.getLogger(BitConverter.class).throwing("BitConverter", "<clinit>", e);
            Utils.die();
        }
    }

    private final static int BUFFER_SIZE = 1024;
    private final static int MAX_SIMILAR_BITS = 512;
    private final static float SYNC_BIT_ACCURACY = .2F;
    private final static Logger LOGGER = Drtd.getLogger(BitConverter.class);

    private final GenericRingBuffer<Boolean> bitBuffer = new GenericRingBuffer<>(Boolean[]::new, false, BUFFER_SIZE);
    private int requiredSyncBits;
    private float[] baudRates;
    private SyncCallback callback;

    private float[] samplesPerBit;
    private boolean lastSample = false;
    private boolean syncing;
    private int sampleRate;
    private float currentSamplesPerBit = 0;
    private int syncBits = 0;

    private final ReceivedLevel receiving = new ReceivedLevel();
    private final ReceivedLevel previouslyReceived = new ReceivedLevel();

    public synchronized void waitForSync() {
        syncing = true;
        currentSamplesPerBit = 0;
        receiving.clear();
        previouslyReceived.clear();
        bitBuffer.clear(false);
    }

    public BitConverter(float baudRate) {
        super(Boolean.class);
        setBaudRates(baudRate);
    }

    public BitConverter(SyncCallback callback, int requiredSyncBits, float... baudRates) {
        super(Boolean.class);
        setBaudRates(baudRates);
        this.callback = callback;
        this.requiredSyncBits = requiredSyncBits;
    }

    public synchronized void setBaudRates(float... baudRates) {
        if (baudRates.length == 0)
            throw new IllegalArgumentException("No baud rates specified!");

        this.baudRates = baudRates;

        if (baudRates.length == 1) {
            requiredSyncBits = 0;
            syncing = false;
            callback = null;
        } else {
            syncing = true;
        }
        recalculate();
    }

    private void recalculate() {
        samplesPerBit = new float[baudRates.length];
        sampleRate = getInputSampleRate();

        float maxSamples = Float.MAX_VALUE;
        for (int i = 0; i < baudRates.length; ++i) {
            float samples = getInputSampleRate() / baudRates[i];
            samplesPerBit[i] = samples;
            maxSamples = Math.min(samples, maxSamples);
        }

        if (samplesPerBit.length == 1)
            currentSamplesPerBit = samplesPerBit[0];
        else
            currentSamplesPerBit = 0;
    }

    @Override
    public int onInit(int calculatedInputSampleRate) {
        recalculate();
        return calculatedInputSampleRate;
    }

    @Override
    public Dimension calculateSize(Graphics2D g) {
        return SIZE;
    }

    @Override
    protected void drawRelative(Graphics2D graphics) {
        graphics.drawImage(IMAGE, 0, 0, null);
    }

    @Override
    protected synchronized Boolean calculate(Boolean sample) {
        if (sample == lastSample) {
            ++receiving.samples;
            return bitBuffer.pop();
        }

        receiving.value = lastSample;
        lastSample = sample;

        if (syncing && samplesPerBit.length > 1) {
            if (currentSamplesPerBit == 0) {
                for (float spb : samplesPerBit) {
                    if (Math.abs(1 - ++receiving.samples / spb) <= SYNC_BIT_ACCURACY) {
                        currentSamplesPerBit = spb;
                        ++syncBits;
                        break;
                    }
                }
            } else {
                int cnt = receiving.bitCount();
                if (cnt == 1) {
                    ++syncBits;
                    if (syncBits == requiredSyncBits) {
                        syncBits = 0;
                        syncing = false;

                        float bauds = getBaudrate();
                        LOGGER.fine("Synced to " + bauds + " baud");
                        if (callback != null)
                            callback.onSyncAquired(currentSamplesPerBit, bauds);
                    }
                } else {
                    currentSamplesPerBit = 0;
                    syncBits = 0;
                }
            }

            receiving.clear();
            return bitBuffer.pop();
        }

        if (currentSamplesPerBit == 0)
            return bitBuffer.pop();

        if (bitBuffer.getCount() == BUFFER_SIZE) {
            LOGGER.warning("Buffer full!");
            return bitBuffer.pop();
        }

        int cnt = receiving.bitCount();
        if (cnt == 0) {
            previouslyReceived.samples += receiving.samples;
            receiving.clear();
            return bitBuffer.pop();
        } else if (cnt >= MAX_SIMILAR_BITS) {
            receiving.clear();
            return bitBuffer.pop();
        }

        previouslyReceived.pushIntoBuffer();
        previouslyReceived.from(receiving);
        receiving.clear();

        return bitBuffer.pop();
    }

    public float getBaudrate() {
        return sampleRate / currentSamplesPerBit;
    }

    private class ReceivedLevel {
        private boolean value = false;
        private int samples = 0;

        private void clear() {
            samples = 0;
            value = false;
        }

        private void from(ReceivedLevel l) {
            samples = l.samples;
            value = l.value;
        }

        private int bitCount() {
            return Math.round(samples / currentSamplesPerBit);
        }

        private void pushIntoBuffer() {
            for (int i = 0; i < bitCount(); ++i)
                bitBuffer.push(value);
        }
    }

    public interface SyncCallback {
        void onSyncAquired(float samplesPerBit, float baudRate);
    }
}

File: src/dsp/BitConverter.cpp
/*
BSD 2-Clause License

Copyright (c) 2020, Till Mayer
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
#include "BitConverter.hpp"
#include <Fl/fl_draw.H>
#include <cmath>

static constexpr const char* bitconverter_xpm[] {
    "40 30 2 1",
    " 	c None",
    ".	c #000000",
    "........................................",
    ".                                      .",
    ".                                      .",
    ".                                      .",
    ".                                      .",
    ".                                      .",
    ".     ....                    ....     .",
    ".        .                    .        .",
    ".        .         ..         .        .",
    ".        .        .  .        .        .",
    ".        .        .  .        .        .",
    ".        .        .  .        .        .",
    ".        .         ..         .        .",
    ".        .   .            .   .        .",
    ".        .  .              .  .        .",
    ".        . .................. .        .",
    ".        .  .              .  .        .",
    ".        .   .            .   .        .",
    ".        .                    .        .",
    ".        .                    .        .",
    ".        .                    .        .",
    ".        .                    .        .",
    ".        .                    .        .",
    ".        ......................        .",
    ".                                      .",
    ".                                      .",
    ".                                      .",
    ".                                      .",
    ".                                      .",
    "........................................"
};

using namespace Dsp;

BitConverter::BitConverter(float baud_rate)
    : BitConverter(0, { baud_rate }) {
}

BitConverter::BitConverter(u16 required_sync_bits, Buffer<float> baud_rates)
    : RefableComponent<bool, bool, BitConverter>("Bit converter")
    , m_required_sync_bits(required_sync_bits)
    , m_baud_rates(std::move(baud_rates)) {
}

SampleRate BitConverter::on_init(SampleRate sample_rate, int&) {
    recalculate_samples_per_bit();
    return sample_rate;
}

Buffer<float> BitConverter::baud_rates() const {
    return m_baud_rates;
}

void BitConverter::wait_for_sync() {
    m_syncing = true;
    m_current_samples_per_bit = 0;
    m_receiving = {};
    m_received_previously = {};
    m_bit_buffer.clear();
}

void BitConverter::set_baud_rates(Buffer<float> rates) {
    assert(rates.size());
    m_baud_rates = std::move(rates);
    recalculate_samples_per_bit();
}

BitConverter& Dsp::BitConverter::ref() {
    return *this;
}

void BitConverter::recalculate_samples_per_bit() {
    assert(m_baud_rates.size());
    m_samples_per_bit = Buffer<float>(m_baud_rates.size());

    for (size_t i = 0; i < m_baud_rates.size(); ++i)
        m_samples_per_bit[i] = input_sample_rate() / m_baud_rates[i];

    if (m_samples_per_bit.size() == 1) {
        m_current_samples_per_bit = m_samples_per_bit[0];
        m_required_sync_bits = 0;
        m_syncing = false;
    } else {
        m_current_samples_per_bit = 0;
        m_syncing = true;
    }
}

bool BitConverter::try_pop_bit_or_abort() {
    if (m_bit_buffer.count())
        return m_bit_buffer.pop();

    GenericComponent::abort_processing();
    return false;
}

Size BitConverter::calculate_size() {
    int width = 0;
    int height = 0;
    fl_measure_pixmap(bitconverter_xpm, width, height);
    return { static_cast<unsigned>(width), static_cast<unsigned>(height) };
}

void BitConverter::draw_at(Point p) {
    fl_draw_pixmap(bitconverter_xpm, p.x(), p.y());
}

float BitConverter::current_baud_rate() const {
    return input_sample_rate() / m_current_samples_per_bit;
}

bool BitConverter::process(bool sample) {
    if (sample == m_last_sample) {
        ++m_receiving.samples;
        return try_pop_bit_or_abort();
    }

    m_receiving.value = m_last_sample;
    m_last_sample = sample;

    if (m_syncing) {
        if (m_current_samples_per_bit == 0) {
            for (auto samples : m_samples_per_bit) {
                if (std::abs(1 - static_cast<float>(m_receiving.samples) / samples) <= sync_bit_accuracy) {
                    m_current_samples_per_bit = samples;
                    m_counted_sync_bits = 1;
                    break;
                }
            }
        } else {
            auto count = m_receiving.bit_count(m_current_samples_per_bit);
            if (count == 1) {
                ++m_counted_sync_bits;
                if (m_counted_sync_bits == m_required_sync_bits) {
                    m_counted_sync_bits = 0;
                    m_syncing = false;
                    float bauds = current_baud_rate();
                    logger().info() << "Synced to " << bauds << " bauds";
                    if (sync_callback)
                        sync_callback({ m_current_samples_per_bit, bauds });
                }
            } else {
                m_current_samples_per_bit = 0;
                m_counted_sync_bits = 0;
            }
        }

        m_receiving = ReceivedSymbol();
        return try_pop_bit_or_abort();
    }

    if (m_current_samples_per_bit == 0)
        return try_pop_bit_or_abort();

    if (m_bit_buffer.count() == buffer_size) {
        logger().warning() << "Buffer full!";
        return try_pop_bit_or_abort();
    }

    auto count = m_receiving.bit_count(m_current_samples_per_bit);
    if (!count) {
        m_received_previously.samples += m_receiving.samples;
        m_receiving = ReceivedSymbol();
        return try_pop_bit_or_abort();
    } else if (count >= max_similar_bits) {
        m_receiving = ReceivedSymbol();
        return try_pop_bit_or_abort();
    }

    for (u16 i = 0; i < m_received_previously.bit_count(m_current_samples_per_bit); ++i)
        m_bit_buffer.push(m_received_previously.value);

    m_received_previously = std::move(m_receiving);
    m_receiving = ReceivedSymbol();

    return try_pop_bit_or_abort();
}

u16 BitConverter::ReceivedSymbol::bit_count(float samples_per_bit) {
    return static_cast<u16>(std::roundf(static_cast<float>(samples) / samples_per_bit));
}
